### PR TITLE
 Eng-8711: fixing kafka importer offset issue

### PR DESCRIPTION
**Pull request by JadenW**

### CODE REVIEW

File: src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java [hunk 1]
```diff
@@ -726,6 +807,9 @@ public boolean commitOffset() {
                             offsetCommitResponse = consumer.commitOffsets(offsetCommitRequest);
                         }
                     }
+                    if ((Short)offsetCommitResponse.errors().get(m_topicAndPartition) == ErrorMapping.NoError()){
+                        info("[STEBUG] " + "Committed offset " + offset + " for " + m_topicAndPartition);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java]: what is STEBUG?

Comment (JadenW) [inline src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java]: Oh, this for debug only. Forget to remove this out..

File: src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java [hunk 2]
```diff
@@ -441,75 +453,148 @@ public void shutdown() {
...
+                    } finally {
+                        if (channel != null) {
+                            channel.disconnect();
+                        }
                     }
                 }
+                if (probeException != null) {
+                    error(probeException, "Failed to query all brokers for the offeset coordinator for " + m_topicAndPartition);
+                }
+                backoffSleep(attempts+1);
```
Comment (akhanzode) [inline src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java]: Why sleep when probeException is not null. You are successful in finding coordinator so break

File: src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java [hunk 3]
```diff
@@ -441,75 +453,148 @@ public void shutdown() {
...
+                                            m_consumerSocketTimeout,
+                                            m_fetchSize, CLIENT_ID
+                                            ));
+                            info("Offset Coordinator for " + m_topicAndPartition + " is " + offsetManager);
+                            closeConsumer(consumer);
+                            probeException = null;
+                            consumer = null;
+                            break;
+                        }
+                        probeException = new KafkaStreamImporterException("Failed to get Offset Coordinator for %s",
```
Comment (akhanzode) [inline src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java]: Create exception only on last attempt.

File: src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java [hunk 4]
```diff
@@ -441,75 +453,148 @@ public void shutdown() {
...
+                                ErrorMapping.exceptionFor(metadataResponse.errorCode()), m_topicAndPartition
+                                );
+                    } catch (Exception e) {
+                        probeException = new KafkaStreamImporterException(
```
Comment (akhanzode) [inline src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java]: Same comment as before. Why create exception.

File: src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java [hunk 5]
```diff
@@ -441,75 +453,148 @@ public void shutdown() {
...
+                error(fault, "unable to fetch earliest offset for " + m_topicAndPartition);
+                rsp = null;
             }
-            return -1;
+            return rsp;
+        }
+
+        public long getLastOffset() {
+
+            final int partition = m_topicAndPartition.partition();
```
Comment (akhanzode) [inline src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java]: Comment this method whats difference between earliest and current and when what is picked.

File: src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java [hunk 6]
```diff
@@ -547,33 +632,51 @@ private int backoffSleep(int fetchFailedCount) {
...
                         getOffsetCoordinator();
-                        m_currentOffset.set(getLastOffset(kafka.api.OffsetRequest.LatestTime()));
+                        long lastOffset = getLastOffset();
+                        m_currentOffset.set(lastOffset);
                         if (m_currentOffset.get() < 0) {
                             sleepCounter = backoffSleep(sleepCounter);
                             info("Latest offset not found for " + m_topicAndPartition + " using earliest offset.");
                             //No latest time available so get earliest known for this consumer group.
-                            m_currentOffset.set(getLastOffset(kafka.api.OffsetRequest.EarliestTime()));
+                            // m_currentOffset.set(getLastOffset(kafka.api.OffsetRequest.EarliestTime()));
```
Comment (akhanzode) [inline src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java]: Dont comment remove code. also log message may be not applicable now?

File: src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java (hunk 2 above)
Comment (JadenW) [inline src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java]: I think we meant to break nested for loop when success. Will fix this.

File: src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java (hunk 6 above)
Comment (JadenW) [inline src/frontend/org/voltdb/importclient/kafka/KafkaStreamImporter.java]: yes, fixed.

**Comment by vtkstef** [comment]
--squash merged

**Pull request closed without merging**